cmd/gateway-gcs.go

```diff
@@ -268,17 +279,67 @@ func newGCSGateway(projectID string) (GatewayLayer, error) {
 		return nil, err
 	}
 
-	return &gcsGateway{
-		client:     client,
-		projectID:  projectID,
-		ctx:        ctx,
-		anonClient: anonClient,
-	}, nil
+	gateway := &gcsGateway{
+		client:       client,
+		projectID:    projectID,
+		ctx:          ctx,
+		anonClient:   anonClient,
+		cleanupEndCh: make(chan struct{}),
+	}
+	// Start background process to cleanup old files in minio.sys.temp
+	go gateway.CleanupGCSMinioPath()
+	return gateway, nil
+}
+
+// Cleanup old files in minio.sys.temp of the given bucket.
+func (l *gcsGateway) CleanupGCSMinioPathBucket(bucket string) {
+	it := l.client.Bucket(bucket).Objects(l.ctx, &storage.Query{Prefix: gcsMinioTempPath, Versions: false})
+	for {
+		attrs, err := it.Next()
+		if err != nil {
+			if err != iterator.Done {
+				errorIf(err, "Object listing error on bucket %s", bucket)
+			}
+			return
+		}
+		if time.Since(attrs.Updated) > gcsMultipartExpiry {
+			// Delete files older than 2 weeks.
+			err := l.client.Bucket(bucket).Object(attrs.Name).Delete(l.ctx)
+			if err != nil {
+				errorIf(err, "Unable to delete the object %s", attrs.Name)
+				return
+			}
+		}
+	}
+}
+
+// Cleanup old files in minio.sys.temp of all buckets.
+func (l *gcsGateway) CleanupGCSMinioPath() {
+	for {
+		it := l.client.Buckets(l.ctx, l.projectID)
+		for {
+			attrs, err := it.Next()
+			if err != nil {
+				if err != iterator.Done {
+					errorIf(err, "Bucket listing error")
+				}
+				break
+			}
+			l.CleanupGCSMinioPathBucket(attrs.Name)
+		}
+		select {
+		case <-l.cleanupEndCh:
```

Comment:
There will be one entry for every part, for all the uploads under a given bucket.
E.g, If there are 3 uploads with each 100 parts uploaded like `bucket/obj1`, `bucket/obj2` and `bucket/obj3`, there would be 303 entries (incl. `gcs.json` for every upload) under `bucket`. I.e, for `N` uploads of objects in a given bucket with `K` parts in each, the total number of entries to be deleted in the bucket would be, `N * (K+1)`. Further, in the GCS go SDK we use, there is no provision to delete multiple objects in a single API request (unlike AWS), we will send API requests equal to number of internal objects to be deleted.

For this process to be quick, one or more of the following should be true,
1. There are very few incomplete uploads during cleanup.
2. There are very few parts in an upload.

I am OK with the current blocking behaviour if you know that the above conditions are likely to be met.